workflow-core/src/main/java/com/squareup/workflow1/Tracing.kt

```diff
@@ -0,0 +1,85 @@
+package com.squareup.workflow1
+
+/**
+ * Tracing facade to install later.
+ *
+ * This follows the static service locator that Timber demonstrates.
+ * See [Timber][https://github.com/JakeWharton/timber/blob/trunk/timber/src/main/java/timber/log/Timber.kt].
+ */
+public class Tracing private constructor() {
+
+  public interface Tracer {
+    public fun beginIdSection(label: () -> WorkflowIdentifier)
+    public fun endIdSection(label: () -> WorkflowIdentifier)
+    public fun beginSection(label: () -> String)
+    public fun endSection(label: () -> String)
+  }
+
+  public companion object : Tracer {
+
+    /**
+     * Note that by default we have nothing installed so nothing gets run! Only install something
+     * in an app you want to profile.
+     */
+    @JvmStatic public fun install(tracer: Tracer) {
+      require(tracer != this) { "Don't install Tracing on itself." }
+      synchronized(tracers) {
+        tracers.add(tracer)
+        tracerArray = tracers.toTypedArray()
+      }
+    }
+
+    @JvmStatic public fun uninstall(tracer: Tracer) {
+      synchronized(tracers) {
+        require(tracers.remove(tracer)) { "Can't uninstall tracer not installed." }
+        tracerArray = tracers.toTypedArray()
+      }
+    }
+
+    @JvmStatic public fun clear() {
+      synchronized(tracers) {
+        tracers.clear()
+        tracerArray = emptyArray()
+      }
+    }
+
+    public inline fun <T> traceId(
+      noinline label: () -> WorkflowIdentifier,
+      crossinline block: () -> T
+    ): T {
+      beginIdSection(label)
+      val result = block()
+      endIdSection(label)
+      return result
+    }
+
+    override fun beginIdSection(label: () -> WorkflowIdentifier) {
+      tracerArray.forEach { it.beginIdSection(label) }
+    }
+
+    override fun endIdSection(label: () -> WorkflowIdentifier) {
+      tracerArray.forEach { it.endIdSection(label) }
+    }
+
+    public inline fun <T> trace(
+      noinline label: () -> String,
+      crossinline block: () -> T
+    ): T {
+      beginSection(label)
+      val result = block()
+      endSection(label)
+      return result
+    }
+
+    override fun beginSection(label: () -> String) {
+      tracerArray.forEach { it.beginSection(label) }
+    }
+
+    override fun endSection(label: () -> String) {
+      tracerArray.forEach { it.endSection(label) }
+    }
+
+    private val tracers = ArrayList<Tracer>()
+    @Volatile private var tracerArray = emptyArray<Tracer>()
```

Comment:
Very confused at first by the need of both an array list and also an array. I think this seems to be related to Volatile here, i.e. the Array is immutable and its ref is volatile so any update to the ref propagates through CPU caches and that means beginSection and endSection don't need any lock. And keeping the list makes it easy to create a new array?

That's definitely a bit surprising, you'll want to document this well.

Reply:
Your logic sounds right - to be honest this was cribbed from Timber's service locator and I didn't think too deeply about it.